Revise SRS (#30)
ggorlen/cs195-project: documentation/SOFTWARE_REQUIREMENTS_SPEC.md

### Phase 1: Basic Gameplay Requirements

#### Game Board

> * The board shall be approximately 500 pixels tall by 600 pixels wide.

Reviewer: Will this work for mobile (or do we even care about mobile -- I vote yes)? Do we need to specify the dimensions? If so, how did we arrive at the 600x500 choice?

Author: The dimension is arbitrary. What would be a better default dimension?

Reviewer: I suggest we avoid committing to some arbitrary values we have no justification for. I think it's better to skip it instead of specify unless we're >90% sure of the value, or we have some sort of rationale to back it up, or we can more clearly state that they're arbitrary.

* The board shall have top view perspective (viewed from above). 

> * Every element in the board shall have a graphical dimensions of 20 pixels by 20 pixels.

Reviewer: We might want to leave this up in the air or allow for responsiveness/mobile view as mentioned above.

Reviewer 2: Then, does that mean the web browser will adjust for the monitor resolution or will it be like steam where we can have a default resolution when a window opens up?

Author: I was just thinking of default dimensions for the graphic designer. But maybe this is not the right document for that. I know that any size of the image can be adjusted  through a stylesheet but it gives the designers an idea of how big our images and shape (square vs rectangular).

Author: ```suggestion
* Every element in the board shall have a square graphical dimension.
```

The game shall be run on any browser that supports WASM (Safari, Chrome, Firefox, Edge, Opera, IOS Safari, Firefox for Android)

### Use Cases

> * The user presses the start button to initiate the game. A prompt will display asking for the user to enter level. A static board elements will be displayed on the screen.

Reviewer: ```suggestion
* The user presses the start button to initiate the game. A prompt will display asking for the user to enter level. A static board element will be displayed on the screen.
```

"Prompt" in web parlance usually means one of those annoying popup dialogs, e.g. `prompt("choose a level")`. Maybe a separate screen showing the levels graphicallly, or a modal, would be better?

* The user can press any directional keys to move around the board. When the user preses UP on the keyboard, the character will move one cell up if the cell is empty or if there is a box and the adjacent cell of the box is empty. The user uses the UP, DOWN, LEFT, RIGHT keys to navigate around until all of the boxes are places on marked goals. This will clear the level.
* When all of the boxes are placed on the goal cells, a congratulatory message will appear.

> * The user will be prompted to move to the next level or jump to a different level.

Reviewer: Maybe automatically moving them directly to the next level and providing a menu option to change levels any time would be best, to avoid decision fatigue and keep the flow going.

Author: I like that. Maybe, we can have a count down that can automatically transition to the next level and also an option to click the NEXT (for eager players) and PLAY AGAIN (for players who want to redo the level.
![Screen Shot 2022-02-27 at 8 50 21 AM](https://user-images.githubusercontent.com/76274300/155895656-8da6e1ef-4a3f-449c-9295-b8018b0ba5fa.png)

Reviewer: I think of puzzle games as relaxing, so I'd prefer a non-countdown -- avoiding presenting it to feel like a race or forced competition/gamification. Maybe it's just an older generation thing ;-).

Author: I was thinking of it as a heads up that the default setting for the game is an "auto-play" so we'll pause for a few seconds before transitioning to the next level and at the same time allowing our users to give them the time to choose to either play the same level or quickly move on to the next level.

Reviewer: Instead of autoplay and a countdown, how about the screen shows a "level complete" message or something, then waits for the user to press a button to play the next level, or returns them to the level selection screen?

Reviewer: We could also make autoplay toggle-able, maybe with the default off. I'll approve the PR and I think we can tweak this as needed later on. One thing to keep in mind is that anything with a timer can make Puppeteer testing a little bit slower/harder/more involved.

Author: Does this restrict js animations?

Reviewer: Yes, in the sense that the extent to which we add animations moves the project closer to the "harder to test" zone we were originally trying to avoid, namely a real-time game with things moving nondeterministically around the screen. We could add some of these things, but the more we add, the more pain it is (maybe not a bad thing, and we don't need to test everything). I'm just erring towards simplicity for starters.

Puppeteer has tools to wait for things to occur, but if everything on the screen updates instantly, we don't need to use them and our tests will be easier to write, faster to run (everything is real-time, so if we have a 3 second animation that we test, that test will take 3 seconds longer to run every time) and less prone to errors that can show up trying to synchronize Puppeteer with the live app.

### Use Cases
* The user presses the start button to initiate the game. A prompt will display asking for the user to enter level. A static world elements will be displayed on the screen.
* The user can press any directional keys to move around the board. When the user preses UP on the keyboard, the character will move one cell up if the cell is empty or if there is a box and the adjacent cell of the box is empty. The user uses the UP, DOWN, LEFT, RIGHT keys to navigate around until all of the boxes are places on marked goals. This will clear the level.

> * When all of the boxes are placed on the goal cells, a congratulatory message.

Author: ```suggestion
* When all of the boxes are placed on the goal cells, a congratulatory message will appear.
```
